### ISSUE TITLE
WinUI MediaPlayerElement

### ISSUE THREAD

**Issue by Fernand-Delavy** [microsoft/microsoft-ui-xaml#5172]
I need MediaPlayer on WinUI or Reunion for Windows

On the page : 
https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.mediaplayerelement?view=winrt-20348

I read : Equivalent WinUI class: Microsoft.UI.Xaml.Controls.MediaPlayerElement.

if using Microsoft.UI.Xaml.Controls.MediaPlayerElement;

But MediaPlayerElement is not present on :

WinUI UWP, WinUI for Desktop , Project Reunion 0.5.7 and 0.8.0 with .NET 5

What is wrong ? How to use MediaPlayerElelemnt ?

**Comment by StephenLPeters** [comment]
MediaPlayerElement isn't currently supported in Winui3,

@MikeHillberg @codendone or @Austin-Lamb to weigh in on the plan for it.

**Comment by codendone** [comment]
See the [WinUI 3.0 Feature Roadmap](https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md#winui-30-feature-roadmap) for the latest status on plans, including **Media Controls**. We currently do not have a specific release timeline for MediaPlayerElement.

**Comment by maxkatz6** [comment]
I wonder if it will include MediaElement or will it be removed as previously deprecated control.

**Comment by DarrylDreiling-Sphero** [comment]
It would be pretty helpful for us to have the MediaPlayerElement  available to WinUI applications.  

We use video content in our educational app to help users learn and complete their school activities, so Media Player support would be very helpful. Rather than sending students out of the app to a browser to view the video content, we’d like to have them view their content inside the app while doing their programming work.

We also believe the MediaPlayerElement  would allow us to swap out our large library of sounds (which are mono, low bit-rate .wav files), for higher fidelity stereo .mp3 files. This would help us reduce our app install size as well.  The Sound Player we are currently using doesn't have a way for us to detect when a sound is done playing, the MediaPlayerElement does.

Overall, better video playback would ensure that our Windows app is competitive with the other versions of our app on iOS, macOS, Android, and ChromeOS, and better ensure the Windows app is a viable solution for our ~200,000 MAU during the school season.

**Comment by qJake** [comment]
I wanted to port my WPF app to WinUI 3 which relies heavily on video playback. This is very disappointing. Please add support for this!

**Comment by erikvanappeldoorn** [comment]
I contribute to an UWP app with background audio streaming (radio). Would be nice to convert the app to WinUI 3, so no media support would be a show stopper.

**Comment by tesar-tech** [comment]
I would like to update my [Video Detail Player](https://www.microsoft.com/en-us/p/video-detail-player/9p34ghb2h88r?activetab=pivot:overviewtab) to WinUI 3 in the near future. This kind of blocks me. The `MediaPlayerElement` in UWP evolved in pretty useful control I have to say, it's shame to leave it behind (But I also understand, it's not easy to port it).

**Comment by agenne** [comment]
We urgently need the Control MediaPlayerElement or MediaElement! The roadmap says it will come in a future release, but we need to know when (Q1, Q2?). We would very much like to switch our UWP app to WinUI3 because we urgently need WebView2, but as long as there is no MediaElement, we cannot make the switch.

**Comment by HermanEldering** [comment]
For me this issue is also problematic since it is not possible to use XAML Islands in WPF with .NET 5/6 and WinUI3 doesn't support MediaPlayerElement. So other than creating/using a third party media control I think the only solutions are to use either .NET Core 3.1 or UWP? Which both are end-of-life.

**Comment by agenne** [comment]
Microsoft, please tell us, will you offer the MediaControls at all? More than half a year is invested in the acrylic features. Who needs that? Would it be possible to know how much longer we have to wait? Unbelievable!

**Comment by stefffdev** [comment]
You can load the video in a WebView2 and use them HTML5 video API via JavaScript interaction.

**Comment by agenne** [comment]
> You can load the video in a WebView2 and use the HTML5 video API via JavaScript interaction.

That's a possibility if playing videos is a marginal feature. We have a digital signage system where this is one of the core features...
In addition, there are problems in memory management with WebView2 that can only be solved with WorkAround's.

**Comment by stefffdev** [comment]
> > For me this is probably also not a solution because I create an overlay using a IBasicVideoEffect.
> 
> Yes, audio and video effects are a must have, at least for me.

Never tried it but maybe you could implement custom effects in JavaScript as well and trigger them from within your .NET app.

**Comment by tesar-tech** [comment]
> Never tried it but maybe you could implement custom effects in JavaScript as well and trigger them from within your .NET app.
Yeah, or we can try to write the whole app in some other technology than c# and xaml. ToTalLy ViAbLe oPtioN... 
"Fck you in particular UWP developers!" No feature parity in platform you invested in, and even not enough feedback!

**Comment by ianier** [comment]
> Yeah, or we can try to write the whole app in some other technology than c# and xaml. ToTalLy ViAbLe oPtioN... "Fck you in particular UWP developers!" No feature parity in platform you invested in, and even not enough feedback!

Indeed. Quoting myself from https://github.com/microsoft/WindowsAppSDK/discussions/1055#discussioncomment-1282924:
"Otherwise, if the cost of moving ahead is rewriting most of the app, I'd rather do it on a platform that shows better commitment to its developers."

**Comment by agenne** [comment]
This is exactly where Microsoft finally has to say honestly if and when the important controls (MediaPlayerElement, InkCanvas and Win2D) will be made available. Not making any statements is very disrespectful to the community.

**Comment by zipswich** [comment]
This is the only remaining showstopper for our primary app that relies on MediaPlayerElement.

**Comment by pwillies** [comment]
We are in a similar situation to many on this thread. We want to move up to WebView2, but our app also has a video/audio playing component. So - WinUI 2.8 (with WebView2) is not released - and with no clear timeframe on release. And WinUI 3.0 has WebView2, but is missing MediaPlayerElement etc. - and no clear timeframe on release.

**Comment by nikolayvpavlov** [comment]
Looking at the number of commits and active contributors draws a dark shadow over my heart.  For such a project to succeed, there must be a big team working actively on it.  Now, WinUI 3 looks like a hobbist project.  MAUI gets more love even in the newsletters, while WinUI 3 is nowhere to be seen.

**Comment by ianier** [comment]
> In all fairness, UWP still works just fine as-is. 

I kind of agree with this statement, as long as you don't rely on too many 3rd party libraries. The feature I'm perhaps hoping the most for is ARM64EC support in .Net Native, which in all fairness, isn't supported in .Net 5/6 either!

**Comment by nikolayvpavlov** [comment]
But really, look at this: [Accouncing Net 7 Preview 1](https://devblogs.microsoft.com/dotnet/announcing-net-7-preview-1/)
There is _not a single word_ about WinUI.  Not a single one.

**Comment by asklar** [comment]
Hey folks,
I've put together a sample for what could be a replacement for `MediaPlayerElement` in WinAppSDK for scoped-down scenarios; it's a prototype at this point, doesn't support fullscreen, I haven't tried it with DRM content, and it does not include the full media transport controls, but maybe even in this state it is useful to the folks on this thread. 
You can find it here: https://github.com/asklar/WinAppSDK-MediaPlayer

PRs welcome :)

**Comment by agenne** [comment]
It's more of a strategic thing. I am not allowed to use any third party technology that is a core functionality of the application. For apps where that is a minor matter, great thing.

**Comment by asklar** [comment]
Hi @agenne,
Disclaimer: I'm a developer in the Windows Developer platform team :)
In order for this control to be part of the built-in WinUI 3 controls., it would likely require significant amounts of additional work to fill in the gaps relative to what the XAML control offers. Even if this was treated as just a stepping stone, it would take work to convert the prototype into a reusable control, integrate it into the existing codebase, etc. so it's "free" as in "free puppy", not as in "free beer" :).

For many apps, those gaps are non-blocking, so they would be able to be unblocked by the code I shared above (maybe they just need a little bit more e.g. a seek bar).

**Comment by MuleaneEve** [comment]
@asklar Thank you for providing the sample. Having something basic that can fill the need of just playing a media file would indeed alleviate a large part of this issue.

Can you explain what the sample is doing? (You could put it directly in your ReadMe)
As far as I can see, it is a regular UWP app...

Also, is it in C++ because it uses Direct3D? (If so, could something like SharpDX or its successor achieve the same result in C#?)

**Comment by HermanEldering** [comment]
I have been thinking of a similar approach to @asklar but instead of referencing ` winrt::Windows::Media::Playback::MediaPlayer` I was thinking of using IBasicVideoEffect. I already have the c# code to work with IBasicVideoEffect, but the drawback is that it probably wouldn't have the timing features (droping frames etc).

I'd probably implement my app with WPF or Avalonia then though...

**Comment by agenne** [comment]
@asklar 
I can well imagine that it is a lot of work. But planning something and then communicating should be quicker. The Windows Developer platform team should tell us when the feature is coming, and if it's coming at all. Currently it says "in one of the future releases". How are you supposed to plan?

**Comment by ianier** [comment]
@HermanEldering: in addition to the timing features, unfortunately using IBasicVideoEffect in C# is painfully slow due to the frequent interop calls, even under .Net Native. 
By looking at @asklar's example I realize that, while awaiting the full-fledged MediaPlayerElement, a MediaPlayerFrameServerPanel control or similar intermediate solution could do the trick for many apps.

**Comment by agenne** [comment]
Agreed, but a CSharp example would be very helpful

**Comment by Hanzalah-Adalan** [comment]
planned for future just like the doomsday..

**Comment by zipswich** [comment]
@asklar Thank you for the effort. We use MPE as follows:

`myMediaPlayerElement.Source = MediaSource.CreateFromMediaStreamSource(myCustomMediaStreamSourceVideo);`

**Comment by asklar** [comment]
@zipswich It seems like that should work fine, since `MediaPlayer.Source` can take the `MediaSource` returned from `CreateFromMediaStreamSource`.

**Comment by zipswich** [comment]
@asklar Thank you for the confirmation. I am using [Uno Platform](https://blogs.windows.com/windowsdeveloper/2021/03/11/build-cross-platform-applications-with-winui-and-uno-platform/). Any pointer to a source to help me integrate your repository with Uno would be greatly appreciated.

**Comment by asklar** [comment]
@zipswich sorry I'm not familiar enough with Uno or its inner workings, I recommend filing an issue in their repo.

**Comment by zipswich** [comment]
@asklar How can this be used by a C# based app? Is there an example? The last time I mixed C++ code with C# was using [SimpleCommunication](https://github.com/microsoft/Windows-universal-samples/tree/main/Samples/SimpleCommunication)

**Comment by agenne** [comment]
There is another new web site for planning winappsdk.
https://portal.productboard.com/winappsdk/1-windows-app-sdk/tabs/1-under-consideration
Could you please all upvote the MediaControls
Thank You in advance

**Comment by PierreHenriKT** [comment]
Hello,

Thanks to @asklar's sample, I managed to publish a [NuGet package that provides a Media Player Element for WinUI 3](https://www.nuget.org/packages/KlearTouch.MediaPlayer.WinUI/). You can find [a sample here](https://github.com/KlearTouch/KlearTouch.MediaPlayer/tree/main/samples).

It is still fairly basic and there are some issues to resolve, so the package is marked as prereleased for now.

This package is designed to be temporary: It uses the `Microsoft.UI` namespace so that, when WinUI introduces its own `MediaPlayerElement`, you should be able to just remove this package, and change nothing else in your code. Although, I'm not sure if that's ok...

The main issue is that the sample app fails to run when adding the `MediaPlayerElement` directly in XAML. So for now, [that part is done in the code-behind](https://github.com/KlearTouch/KlearTouch.MediaPlayer/blob/2ed0e62063a6145db68289faee3ddc9042d26ae1/samples/KlearTouchApp.MediaPlayer.WinUI/MainWindow.xaml.cs#L11).

If you wish to discuss this package in-depth, please [create a new issue in its repo](https://github.com/KlearTouch/KlearTouch.MediaPlayer/issues) (let's not side-track the goal of this issue).

**Comment by emceelovin** [comment]
this is great. the ability to actually create a custom stream source is the problem with EVERY MediaPlayer/Element in windows. EXCEPT the UWP implementation. Which offers in/out stream configurations(paramount IMO for VoIP video chat etc). A/V is paramount in applications. Microsoft always seems to ditch necessary things and work on other stuff that's useless. Great job man!

**Comment by castorix** [comment]
> Agreed, but a CSharp example would be very helpful

Another way to play videos is to embed the WMP control
I tested with VB.NET/WinUI 3, with ATL to embed it ("AtlAxWin" class), but I needed a Win32 Region so that it appears... 

![WinUI3_WMP](https://user-images.githubusercontent.com/22345506/161698676-679cbcd3-57f6-43eb-8277-555ceba1499b.jpg)

**Comment by brabebhin** [comment]
Hi @asklar 

Your sample will crash when attempting to do subtitles.
You can check with this URL:

https://mediaplatstorage1.blob.core.windows.net/windows-universal-samples-media/elephantsdream-clip-h264_sd-aac_eng-aac_spa-aac_eng_commentary-srt_eng-srt_por-srt_swe.mkv

**Comment by castorix** [comment]
I just tested the [Media Engine](https://docs.microsoft.com/en-us/windows/win32/api/mfmediaengine/nn-mfmediaengine-imfmediaengine) and it seems to work very fine
I tested in C# on Windows 10, in Frame-server mode with DXGI or in classic Rendering mode, 
It is atm the best Video Player that I could test with WinUI 3 (I tested WMP, Direct Show, ...)

![WinUI3_MediaEngine](https://user-images.githubusercontent.com/22345506/168442541-25599ccd-e837-4f26-81b6-881667904d9c.jpg)

**Comment by MuleaneEve** [comment]
@castorix Have you tried this [simple MediaPlayerElement](https://www.nuget.org/packages/KlearTouch.MediaPlayer.WinUI/)?

It doesn't have all the features built-in yet, but I think it is the best solution in the long term.

**Comment by castorix** [comment]
> @castorix Have you tried this [simple MediaPlayerElement](https://www.nuget.org/packages/KlearTouch.MediaPlayer.WinUI/)?

Yes, I had tried, but it has the known resizing problem of the _SwapChainPanel_ when scale > 100%
By using _CreateSwapChainForHwnd_ , then resizing the target surface from any XAML control, I don't have this problem with the Media Engine or any Direct2D content

**Comment by PierreHenriKT** [comment]
That issue was resolved a couple of weeks ago. Feel free to test it.

**Comment by castorix** [comment]
> That issue was resolved a couple of weeks ago. Feel free to test it.
> (Edit: There is still an issue when the window is small, though...)

Yes, I re-tested and have (a bit of) flickering on resizing with small size (and also some crashes on longer videos (Windows 10 21H1), on seeking for example)

**Comment by nikolayvpavlov** [comment]
@castorix , can you please provide a sample how to use this Media Engine from C#?

**Comment by brabebhin** [comment]
Media Engine is essentially the engine behind the UWP MediaPlayer class. It cannot be trivially used in C#. 
The problem is that it doesn't support subtitles.

**Comment by castorix** [comment]
> The problem is that it doesn't support subtitles.

I did not test all ME interfaces, but I could draw a text on the DXGI surface, just after [IMFMediaEngine::TransferVideoFrame](https://docs.microsoft.com/en-us/windows/win32/api/mfmediaengine/nf-mfmediaengine-imfmediaengine-transfervideoframe)
So subtitles could theorically be drawn, with the right timing...
(I will post a C# code sample soon, but I'm actually doing quick tests with "dirty" code... (I'm adding buttons atm, not ready yet...))
Test (with GDI on [IDXGISurface1](https://docs.microsoft.com/en-us/windows/win32/api/dxgi/nn-dxgi-idxgisurface1)) : 
![WinUI3_MediaEngine_Text](https://user-images.githubusercontent.com/22345506/168891774-ab0a38e1-5011-4f2f-99f2-9a68cddb0def.jpg)

**Comment by brabebhin** [comment]
Basically, yes.
If you use the MediaPlayer class in frame server mode, it will tell you when to render subtitles.

**Comment by nikolayvpavlov** [comment]
@PierreHenriKT , is there a way to set the current position of the media in the component?

**Comment by PierreHenriKT** [comment]
It is a replica of the [UWP MediaPlayerElement](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Controls.MediaPlayerElement), so you can use (most of) the same features. [For the position, refer to this documentation](https://docs.microsoft.com/en-us/windows/uwp/audio-video-camera/play-audio-and-video-with-mediaplayer#playback-session).

**Comment by castorix** [comment]
> Basically, yes. If you use the MediaPlayer class in frame server mode, it will tell you when to render subtitles.

Apparently, subtitles seem to be managed with [IMFTimedText ](https://docs.microsoft.com/en-us/windows/win32/api/mfmediaengine/nn-mfmediaengine-imftimedtext) and associated interfaces
I will test asap...

**Comment by castorix** [comment]
> @castorix , can you please provide a sample how to use this Media Engine from C#?

I just uploaded a sample : [WinUI3_MediaEngine](https://github.com/castorix/WinUI3_MediaEngine)
I tested various things, like Frame capture, which were more complicated that I thought... so , there are probably a few bugs in this test...
I added Subtitles, which work, but partially : I did not find a way to get styles (italic, bold, ...) with
IMFTimedText* interfaces (I tested for example with Terminator 2 subtitles, which should be in italic at beginning...)

**Comment by brabebhin** [comment]
I also have a semi working approach to subtitles that use the winRT interfaces, but positioning is really difficult to get working correctly with it comes to all the wanky stuff that SSA supports. 

I still very much hope that Microsoft sees the light and releases a proper Mediaplayerelement.

**Comment by eshvatskyi** [comment]
I'm using LibVLC + WriteableBitmap, as a temp solution

**Comment by castorix** [comment]
> I just uploaded a sample : [WinUI3_MediaEngine](https://github.com/castorix/WinUI3_MediaEngine) 

I had to modify the SwapChain and the method for FullScreen because of update 1.1.0 (with the new black/white background added, apparently for flickering problem...)

**Comment by emceelovin** [comment]
> I'm using LibVLC + WriteableBitmap, as a temp solution

I downloaded LibVLC also as the base for creating a media player inside my application as well. Haven't had much time to finish it though. Seems like a solid solution. When I was reading the docs, it looks like you can pass it an HWND(for Win32) and other native window targets and it'll render the media for you. Awesome IMO.

**Comment by alfamizar** [comment]
@emceelovin, @eshvatskyi
Hi, can you guys show ho do you extract bytes that are writting than in to the WriteableBitmap, please?

**Comment by PierreHenriKT** [comment]
Apparently, [MediaPlayerElement was just added to WinUI 3 (in the Windows App SDK version 1.2 Preview 1)](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/preview-channel#winui-3).

That's a pleasant surprise!

**Comment by alfamizar** [comment]
Thank you very much for reply!Really great news.I think this will unblock vlc and maui for video playback support soon.7 сент. 2022 г. 14:35 пользователь Pierre Henri ***@***.***> написал:
Apparently, MediaPlayerElement was just added to WinUI 3 (in the Windows App SDK version 1.2 Preview 1).
That's a pleasant surprise!

**Comment by codendone** [comment]
@PierreHenriKT If you haven't done so, please log a new issue for the crashes you're experiencing with some of the TransportControls. Thanks!

Regarding FullWindow and CompactOverlay, these will not be supported in 1.2. Those buttons and APIs have been removed until they are implemented in a future release.

**Comment by PierreHenriKT** [comment]
@codendone [Issue logged](https://github.com/microsoft/microsoft-ui-xaml/issues/7702).

Thanks for the explanation for FullWindow and CompactOverlay.

**Comment by JPDelprat** [comment]
Hello,

   I am successful in readin a mpeg-dash stream with play ready protection with UWP code (MediaPlayerElement / AdaptiveMediaSource).

With the same code using WinUI3 (last experimental version), I get a "SourceNotSupported" exception.

Are there any differences between WinUI3 and UWP implementations ? Do I miss something ?

Thanks

**Comment by brabebhin** [comment]
It is probably just bugged.
There were a lot of bugs last time I checked.

**Comment by PierreHenriKT** [comment]
With [the release of the Win App SDK 1.2](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#version-12), I think we can close this issue :)

**Comment by Fernand-Delavy** [comment]
Yes you can close this issue, thanks

**Comment by PierreHenriKT** [comment]
@Fernand-Delavy You are the one who opened this issue, so you can close it : There is a button for that next to the "Comment" button..

**Comment by Hanzalah-Adalan** [comment]
OMG finally :)
